mgr/iostat: implement 'ceph iostat' as a mgr plugin

mogeb opened the pull request:
Requires pg_dump to include the PGMap::Incremental object, which holds
the IO activity recently completed by the OSDs.

Signed-off-by: Mohamad Gebai <mgebai@suse.com>

mogeb commented:
IO statistics given by `ceph -s`:

```
$> while true; do ./bin/ceph -s 2>&1 | grep client; sleep 1; done
    client:   21488 kB/s wr, 0 op/s rd, 10 op/s wr
    client:   21488 kB/s wr, 0 op/s rd, 10 op/s wr
    client:   21489 kB/s wr, 0 op/s rd, 10 op/s wr
    client:   74705 kB/s wr, 0 op/s rd, 36 op/s wr
    client:   74709 kB/s wr, 0 op/s rd, 36 op/s wr
    client:   74712 kB/s wr, 0 op/s rd, 36 op/s wr
    client:   74712 kB/s wr, 0 op/s rd, 36 op/s wr
    client:   120 MB/s wr, 0 op/s rd, 60 op/s wr
    client:   120 MB/s wr, 0 op/s rd, 60 op/s wr
    client:   102005 kB/s wr, 0 op/s rd, 49 op/s wr
    client:   148 MB/s wr, 0 op/s rd, 74 op/s wr
```

`ceph iostat` running simultaneously:

```
$> ./bin/ceph iostat
wr: 21488 kB/s, rd: 0 kB/s, iops: 10
wr: 21489 kB/s, rd: 0 kB/s, iops: 10
wr: 21489 kB/s, rd: 0 kB/s, iops: 10
wr: 74705 kB/s, rd: 0 kB/s, iops: 36
wr: 74709 kB/s, rd: 0 kB/s, iops: 36
wr: 74709 kB/s, rd: 0 kB/s, iops: 36
wr: 74712 kB/s, rd: 0 kB/s, iops: 36
wr: 123501 kB/s, rd: 0 kB/s, iops: 60
wr: 123501 kB/s, rd: 0 kB/s, iops: 60
wr: 102005 kB/s, rd: 0 kB/s, iops: 49
wr: 152495 kB/s, rd: 0 kB/s, iops: 74
```

mogeb commented on src/ceph.in:
```diff
@@ -1005,6 +1006,16 @@ def main():
+    if len(childargs) > 0 and childargs[0] == 'iostat':
+        def call_iostat():
+            while 1:
+                subprocess.call([sys.argv[0], 'mgr', 'iostat'])
```
As of now, it doesn't seem like there's a way to keep an open connection between a mgr plugin and the client. The alternative, which is implemented here, is having the client invoking the `iostat` plugin periodically.

Liuchang0812 replied:
@badone added a blocking mode in `scrub` command. https://github.com/ceph/ceph/pull/19793
 maybe we could do this in same way.

mogeb replied:
@Liuchang0812 Thanks for the comments. The issue is that (to my knowleddge) there's no way for the plugin to continuously communicate with the client. Adding a blocking mode for the client wouldn't solve that.

Liuchang0812 commented on src/mon/PGMap.cc:
```diff
@@ -1367,6 +1367,8 @@ void PGMap::dump(Formatter *f) const
   dump_pg_stats(f, false);
   dump_pool_stats(f);
   dump_osd_stats(f);
+  pg_sum_delta.dump(f);
```
we already have `PGMapDigest::overall_client_io_rate_summary` function, we could achieve `iostats` command via adding a python_get('client_io`) interface. likes https://github.com/ceph/ceph/pull/19985/commits/6272c2636425d98b3194828157366527977277bb

mogeb replied:
yes, `pg_sum_delta` (dumped here) is what is indeed used by `PGMapDigest::overall_client_io_rate_summary`. I added a 'io_rate' as a way to fetch it from a mgr plugin.

mogeb commented:
Implements https://trello.com/c/fe9Dk0JF/281-ceph-iostat

jcsp commented:
I see three reasonable-ish options here:

1. (What this PR currently does) Add special cases to the CLI client, like we have for "ceph -w" and "ceph daemonperf"
2. Extend the blocking command mode to enable sending interim output.  Probably quite awkward to pass all the way up the python stack with callbacks etc for the interim output.
3. Extend the command description format to identify specific commands as "polling" so that the CLI knows to block and call the command repeatedly until the user ctrl-c's it.

I think my instinct is for option 3, anyone else?

Liuchang0812 commented:
I think my instinct is for option 3, anyone else?

+1

mogeb commented:
@jcsp In option 3, there won't really be any polling though, correct? There will still be repeated calls from the client, just like in this PR. In any case, I agree with you for option 3 - I'll update this PR accordingly.

jcsp commented:
@mogeb right, the basic behaviour (sending an MCommand repeatedly) would be the same, it would just be a more generic mechanism so that the CLI doesn't have to be hardcoded with the "magic" commands.

I guess you'll probably want a convention for passing a "first=true" or similar argument, so that the server can generate the table header on the first call, then just output table lines later.  Perhaps also have the CLI send the terminal width as a parameter of the command too.

mogeb commented:
@jcsp I was hoping to keep the formatting in the client. Maybe the plugin can expose a `get header` command, which the client calls when needed (on the first call, and when the header is out of sight). But, as you said, the terminal width would still need to be sent.

What do you think about formatting the returned values in JSON, and having the client unpack it and print it however it wants? This adds more flexibility too, so one can easily fetch the current throughput/iops. Or is this extra conversion something we want to avoid?

jcsp commented:
The trouble is that the client doesn't know anything about the data it's printing.  For example, some plugins might want to selectively drop columns (like we do with "daemonperf") depending on the width of the terminal, and the client doesn't know the rules for that.  Colour coding is also simpler if the plugin can just output its colourized lines rather than having to invent a scheme for the plugin to hint to the client about colours.

mogeb commented:
@jcsp Ah, I understand - you want to get completely rid of all command-specific code on the client. That part went over my head. Ok, I'll update this PR then.

mogeb commented:
@jcsp Ok, here area a few points regarding option 3:

1. `daemonperf` calls `perf dump` on the admin socket, and the results are in JSON, so there isn't really much consolidation we can do

2. The log watch (`-w`) is only invoked once, but it keeps an open connection with the monitor (which is what I was looking to do initially with `iostat`, but we can't keep an open connection with a plugin). Its behavior is different from `iostat`, no consolidation possible here either

3. At the moment, there are no other `ceph` commands that have the same behavior as `iostat`, but I think we can expect more `mgr` plugins commands to have this behavior, which leads to the final point:

4. We currently don't have any command description format we can extend to add the `polling` flag. All subcommands are hardcoded in the monstrous ceph.in file. We could add a command description format but, IMO, it would be better to just refactor that file completely at this point, maybe using python's `ArgumentParser`'s `sub_parsers`. I suggest we start by getting `iostat` in, then we can refactor `ceph.in` independently (or vice versa).

If that makes sense, I'll go ahead and update the output of iostat (table header, etc.) as we discussed.

jcsp commented:
> We currently don't have any command description format we can extend to add the polling flag. All subcommands are hardcoded in the monstrous ceph.in file. We could add a command description format but, IMO, it would be better to just refactor that file completely at this point, maybe using python's ArgumentParser's sub_parsers. I suggest we start by getting iostat in, then we can refactor ceph.in independently (or vice versa).

Fortunately, the commands are definitely not hardcoded in ceph.in (although a few special ones are).

The command descriptions are present in the COMMANDS member of a python module, or in mon/MonCommands.h for the C++ ones.  These are passed to clients (such as the CLI) by the special "get_command_descriptions" command, and interpreted by parse_json_funcsigs and similar code in src/pybind/ceph_argparse.py

mogeb commented:
@jcsp I've updated (and rebased) the PR. I've added a `poll` flag that can be defined either in a mgr module, or in a MonCommand.h. What are your thoughts on this version?

jecluis commented on src/ceph.in:
```diff
@@ -515,6 +517,42 @@ else:
+                print(outs, file=sys.stderr)
+            sleep(1)
+    else:
+        ret, outbuf, outs = json_command(cluster_handle, target=target, argdict=valid_dict,
```
This could be something like

> while True:
> // run the command
>  if 'poll' in valid_dict and valid_dict['poll']:
>    break

no?

mogeb replied:
Sure, that's clearer. Added.

jecluis commented on src/pybind/ceph_argparse.py:
```diff
@@ -24,7 +24,7 @@
 
 FLAG_MGR = 8   # command is intended for mgr
-
+FLAG_POLL = 16
```
mind adding a brief comment telling us where the magic number comes from (i.e., mon/MonCommands.h's command flag)? :)

mogeb replied:
Right :) Done.

jecluis commented on src/ceph.in:
```diff
@@ -515,6 +517,42 @@ else:
+                print(outbuf)
+            if outs:
+                print(outs, file=sys.stderr)
+            sleep(1)
```
Would be nice if the polling interval was configurable, defaulting to some value (may it be 1 or something else). Maybe an option passed to the cli tool?

mogeb replied:
Right, I was thinking of having it in the command description itself, but I think a flag to the cli tool is just as valid and definitely easier. Done.

jcsp commented:
I like this.  Since we're extending the command definition language, let's take the opportunity to make it futureproof and add a way for polling commands to supply a title and have some knowledge of the terminal size to do neat formatting (I'm thinking of commands that have a more table-like format).

Here's what I think we should do for futureproofing:
- CLI client always pass a "title" boolean to the first time it calls (this could later be extended to do periodic title output depending on terminal height)
- pass the terminal width as a command argument (see Termsize in ceph_daemon.py for how to get it).
- pass a boolean for whether color should be used (see Daemonwatcher.supports_color)

We can do that in a followup PR as long as we land it before Mimic, or do it in this PR: what do you think @mogeb ?

mogeb commented:
@jcsp yeah, I had kept all the rendering part of iostat till the end as this RFC was mostly for the approach :) Passing those arguments as you're suggesting makes a lot of sense. I'll follow this up in a separate PR, definitely before the feature freeze of mimic.

jcsp commented:
OK -- we're almost there, just a couple of things noticed when testing:

1. currently it continues to poll even if there's an error, so if I e.g. try using it without the module enabled then I get repetitive output like this:
```
Error: 95 ENOTSUP
Module 'iostat' is not enabled (required by command 'iostat'): use `ceph mgr module enable iostat` to enable it
Error: 95 ENOTSUP
Module 'iostat' is not enabled (required by command 'iostat'): use `ceph mgr module enable iostat` to enable it
```

I guess the poller should just drop out if it gets any error other than EAGAIN?

2. When doing a ctrl-c you get an exception dump like this:
```
^CTraceback (most recent call last):
  File "bin/ceph", line 1195, in <module>
    retval = main()
  File "bin/ceph", line 1126, in main
    verbose)
  File "bin/ceph", line 577, in new_style_command
    ret, outbuf, outs = do_command(parsed_args, target, cmdargs, sigdict, inbuf, verbose)
  File "bin/ceph", line 555, in do_command
    sleep(parsed_args.period)
KeyboardInterrupt
```

Since ctrl-c is the normal way to end the command, I suggest we have a try/except around the sleep, and on KeyboardInterrupt just do an exit(0)

tchaikov commented on src/pybind/mgr/iostat/module.py:
```diff
@@ -0,0 +1,35 @@
+                wr = int(r['pg_stats_delta']['stat_sum']['num_write_kb']) / stamp_delta
+                ops = ( int(r['pg_stats_delta']['stat_sum']['num_write']) + int(r['pg_stats_delta']['stat_sum']['num_read']) ) / stamp_delta
+
+        ret = "wr: {0} kB/s, rd: {1} kB/s, iops: {2}".format(str(int(wr)), str(int(rd)), str(int(ops)))
```
nit, no need to convert rd, wr, and ops to str explicitly. `str()` will be called automatically by `.format()`.

mogeb commented:
@jcsp @tchaikov done. Next up: fixing the output.

PS: By looking at some of the qa code, it seems that the self-test commands are ran automatically. I just want to confirm that this really is the case, and that I don't have to add an invocation to iostat's `self-test` somewhere..

jcsp commented:
> PS: By looking at some of the qa code, it seems that the self-test commands are ran automatically. I just want to confirm that this really is the case, and that I don't have to add an invocation to iostat's self-test somewhere..

Adding the test_iostat method to the exististing TestModuleSelfTest (as you have already done) is sufficient.

LenzGr commented:
Does this PR address the requirements listed on [this Trello card](https://trello.com/c/f2vsUZ39/58-ceph-iostat-pgperf-pool-command)?

mogeb commented:
@LenzGr my first comment on this PR references this card: https://trello.com/c/fe9Dk0JF/281-ceph-iostat
But I assume it's a duplicate (or a subset) of the one you're talking about.

LenzGr commented:
@mogeb thanks - acoording to @liewegas , the usability Trello board was something that was set up for the Luminous release and may be out of date in some places. The card on that board provided some more context than the one you referenced, in particular listing the following information that should be displayed: "read/write ios, read/write bytes, recovery objects or bytes/sec"

jcsp commented:
This is looking pretty complete, is it waiting for anything else @mogeb ?

tchaikov commented:
retest this please

mogeb commented:
@jcsp not waiting for anything other than a couple of lgtms. I'm leaving China next week and I'll start working on the follow-up right after.

mogeb commented:
@LenzGr thanks for the pointer. As discussed on this PR, I'll follow this up with another PR with a proper mechanism for the display.

LenzGr commented:
Any news on this one? Is it good to be merged?

mogeb commented:
FWIW, I started working on having a better output for this command (I'm fine with waiting and merging everything as a single PR, or having separate PRs).

tchaikov commented on src/pybind/mgr/iostat/__init__.py:
```diff
@@ -0,0 +1 @@
+from module import *
```
should be 
```py
from .module import *
```

or better off
```
from module import Module
```

as 

> Explicit is better than implicit.

=)

tchaikov commented on src/pybind/mgr/iostat/__init__.py:
```diff
@@ -0,0 +1 @@
+from module import Module
```
@mogeb should be `from .module import Module`

tchaikov commented:
@mogeb could you highlight the changes you made in the latest version? for instance, how did you fix "this change breaks "ceph pg 1.0 query"?

mogeb commented on src/pybind/ceph_argparse.py:
```diff
@@ -999,6 +1001,9 @@ def validate(args, signature, flags=0, partial=False):
         d['target'] = ('mgr','')
 
+    if flags & FLAG_POLL:
+        d['poll'] = True
```
@tchaikov The problem was that `ceph.in` looks at `len(valid_dict)` (ie the number of parameters) to decide whether it's a `pg` command or not. See [here](https://github.com/ceph/ceph/blob/master/src/pybind/ceph_argparse.py#L1164). Setting `valid_dict['poll'] = False` (previously done here) broke that logic. The fix is to keep `valid_dict['poll']` unset, unless the command server-side expects it. This is OK because the length of `valid_dict` isn't used for the plugin. There's already a check that `'poll' in valid_dict` whenever we want to access it, so having it unset is not an issue.

tchaikov replied:
ahh, thanks for the explanation! actually, i was debugging this yesterday.

mogeb commented:
@tchaikov sorry, I hesitated to do that :) I added a comment that explains the problem.

tchaikov commented on src/ceph.in:
```diff
@@ -515,6 +520,44 @@ else:
+                    file=sys.stderr)
+            break
+        if outbuf:
+            print(outbuf)
```
if `outbuf` is human-readable content, we should `print(outbuf.decode('utf-8'))`, this is important in py3.

tchaikov commented on src/ceph.in:
```diff
@@ -515,6 +520,44 @@ else:
+    if valid_dict:
+        if parsed_args.output_format:
+            valid_dict['format'] = parsed_args.output_format
+
```
please drop this empty line.

tchaikov commented on src/ceph.in:
```diff
@@ -515,6 +520,44 @@ else:
+    while True:
+        ret, outbuf, outs = json_command(cluster_handle, target=target, argdict=valid_dict,
+                inbuf=inbuf)
+
```
please drop this empty line.

tchaikov commented on src/ceph.in:
```diff
@@ -515,6 +520,44 @@ else:
+        if 'poll' not in valid_dict or not valid_dict['poll']:
+            # Don't print here if it's not a polling command
+            break
+
```
please drop this empty line.

tchaikov commented on src/mgr/PyModule.cc:
```diff
@@ -384,6 +384,17 @@ int PyModule::load_commands()
+    item.polling = false;
+    PyObject *pPoll = PyDict_GetItemString(command, "poll");
+    if (pPoll) {
+      std::string polling = PyString_AsString(pPoll);
```
could just use `boost::iequals(polling, "true")`

tchaikov commented on src/pybind/ceph_argparse.py:
```diff
@@ -787,6 +788,7 @@ def parse_json_funcsigs(s, consumer):
         cmd['sig'] = parse_funcsig(cmd['sig'])
         # just take everything else as given
         sigdict[cmdtag] = cmd
+
```
drop this empty  line.

mogeb commented:
@tchaikov sorry for the added iteration. Should be all done.

tchaikov commented:
no worries. thanks for your persistence!

tchaikov commented:
- http://pulpito.ceph.com/kchai-2018-04-14_03:18:44-rados-wip-kefu-testing-2018-04-14-0918-distro-basic-mira/
- http://pulpito.ceph.com/kchai-2018-04-15_01:40:50-rados-wip-kefu-testing-2018-04-14-0918-distro-basic-smithi/
- http://pulpito.ceph.com/kchai-2018-04-15_13:37:16-rados-wip-kefu-testing-2018-04-14-2138-distro-basic-mira/

the failure is addressed by #21433
